File: java-client/src/main/java/com/couchbase/client/java/manager/eventing/EventingFunction.java
/*
 * Copyright 2021 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.java.manager.eventing;

import com.couchbase.client.core.annotation.Stability;

import java.util.ArrayList;
import java.util.List;

import static com.couchbase.client.core.util.Validators.notNull;
import static com.couchbase.client.core.util.Validators.notNullOrEmpty;

/**
 * An immutable representation of the {@link EventingFunction} stored or to be stored on the server.
 */
@Stability.Uncommitted
public class EventingFunction {

  private final String name;
  private final String code;
  private final EventingFunctionKeyspace sourceKeyspace;
  private final EventingFunctionKeyspace metadataKeyspace;
  private final EventingFunctionSettings settings;
  private final boolean enforceSchema;
  private final String version;
  private final int handlerUuid;
  private final String functionInstanceId;
  private final List<EventingFunctionBucketBinding> bucketBindings;
  private final List<EventingFunctionUrlBinding> urlBindings;
  private final List<EventingFunctionConstantBinding> constantBindings;

  /**
   * Creates a new {@link EventingFunction} with the minimal required properties.
   *
   * @param name the name of the function.
   * @param code the code body of the function.
   * @param sourceKeyspace the keyspace from where the source data is coming from.
   * @param metadataKeyspace the keyspace where the function metadata is stored.
   * @return the created {@link EventingFunction}.
   */
  public static EventingFunction create(String name, String code, EventingFunctionKeyspace sourceKeyspace,
                                EventingFunctionKeyspace metadataKeyspace) {
    return builder(name, code, sourceKeyspace, metadataKeyspace).build();
  }

  /**
   * Creates a {@link Builder} that can be used to further customize the {@link EventingFunction} beyond the defaults.
   *
   * @param name the name of the function.
   * @param code the code body of the function.
   * @param sourceKeyspace the keyspace from where the source data is coming from.
   * @param metadataKeyspace the keyspace where the function metadata is stored.
   * @return the builder to customize.
   */
  public static Builder builder(String name, String code, EventingFunctionKeyspace sourceKeyspace,
                                EventingFunctionKeyspace metadataKeyspace) {
    return new Builder(name, code, sourceKeyspace, metadataKeyspace);
  }

  private EventingFunction(Builder builder) {
    this.settings = builder.settings;
    this.name = builder.name;
    this.code = builder.code;
    this.sourceKeyspace = builder.sourceKeyspace;
    this.metadataKeyspace = builder.metadataKeyspace;
    this.bucketBindings = builder.bucketBindings;
    this.urlBindings = builder.urlBindings;
    this.constantBindings = builder.constantBindings;
    this.enforceSchema = builder.enforceSchema;
    this.handlerUuid = builder.handlerUuid;
    this.functionInstanceId = builder.functionInstanceId;
    this.version = builder.version;
  }

  /**
   * The name of the function.
   */
  public String name() {
    return name;
  }

  /**
   * The actual javascript source code of the function.
   */
  public String code() {
    return code;
  }

  /**
   * The source keyspace where the actual data is accessed from.
   */
  public EventingFunctionKeyspace sourceKeyspace() {
    return sourceKeyspace;
  }

  /**
   * The keyspace where eventing stores the metadata for the function.
   */
  public EventingFunctionKeyspace metadataKeyspace() {
    return metadataKeyspace;
  }

  /**
   * The version of the function.
   */
  public String version() {
    return version;
  }

  /**
   * True if the schema should be enforced.
   */
  public boolean enforceSchema() {
    return enforceSchema;
  }

  /**
   * The function UUID, provided by the server.
   */
  public long handlerUuid() {
    return handlerUuid;
  }

  /**
   * The function instance ID, provided by the server.
   */
  public String functionInstanceId() {
    return functionInstanceId;
  }

  /**
   * The custom function settings applied.
   */
  public EventingFunctionSettings settings() {
    return settings;
  }

  /**
   * The bucket bindings for the function.
   */
  public List<EventingFunctionBucketBinding> bucketBindings() {
    return bucketBindings;
  }

  /**
   * The URL bindings for the function.
   */
  public List<EventingFunctionUrlBinding> urlBindings() {
    return urlBindings;
  }

  /**
   * The constant bindings for the function.
   */
  public List<EventingFunctionConstantBinding> constantBindings() {
    return constantBindings;
  }

  /**
   * This builder allows to customize the properties of the eventing function.
   */
  public static class Builder {

    private final String name;
    private final String code;
    private final EventingFunctionKeyspace sourceKeyspace;
    private final EventingFunctionKeyspace metadataKeyspace;

    private EventingFunctionSettings settings = EventingFunctionSettings.create();
    private boolean enforceSchema = false;
    private List<EventingFunctionBucketBinding> bucketBindings = new ArrayList<>();
    private List<EventingFunctionUrlBinding> urlBindings = new ArrayList<>();
    private List<EventingFunctionConstantBinding> constantBindings = new ArrayList<>();

    private String version;
    private int handlerUuid;
    private String functionInstanceId;

    private Builder(String name, String code, EventingFunctionKeyspace sourceKeyspace,
                    EventingFunctionKeyspace metadataKeyspace) {
      this.name = notNullOrEmpty(name, "Name");
      this.code = notNullOrEmpty(code, "Code");
      this.sourceKeyspace = notNull(sourceKeyspace, "SourceKeyspace");
      this.metadataKeyspace = notNull(metadataKeyspace, "MetadataKeyspace");
    }

    /**
     * Sets the URL bindings for the function.
     *
     * @param urlBindings the URL bindings for the function.
     * @return the {@link Builder} for chaining purposes.
     */
    public Builder urlBindings(List<EventingFunctionUrlBinding> urlBindings) {
      this.urlBindings = new ArrayList<>(notNull(urlBindings, "EventingFunctionUrlBinding"));
      return this;
    }

    /**
     * Sets the bucket bindings for the function.
     *
     * @param bucketBindings the bucket bindings for the function.
     * @return the {@link Builder} for chaining purposes.
     */
    public Builder bucketBindings(List<EventingFunctionBucketBinding> bucketBindings) {
      this.bucketBindings = new ArrayList<>(notNull(bucketBindings, "EventingFunctionBucketBinding"));
      return this;
    }

    /**
     * Sets the constant bindings for the function.
     *
     * @param constantBindings the constant bindings for the function.
     * @return the {@link Builder} for chaining purposes.
     */
    public Builder constantBindings(List<EventingFunctionConstantBinding> constantBindings) {
      this.constantBindings = new ArrayList<>(notNull(constantBindings, "EventingFunctionConstantBinding"));
      return this;
    }

    /**
     * Set to true if the schema schould be enforced.
     *
     * @param enforceSchema true if it should be enforced.
     * @return the {@link Builder} for chaining purposes.
     */
    public Builder enforceSchema(boolean enforceSchema) {
      this.enforceSchema = enforceSchema;
      return this;
    }

    /**
     * Sets various function settings to customize the runtime behavior.
     *
     * @param settings the settings to apply to the function.
     * @return the {@link Builder} for chaining purposes.
     */
    public Builder settings(EventingFunctionSettings settings) {
      this.settings = settings;
      return this;
    }

    /**
     * (internal) Sets the version of the function, set when loading from the server.
     *
     * @param version the version provided from the server.
     * @return the {@link Builder} for chaining purposes.
     */
    Builder version(String version) {
      this.version = version;
      return this;
    }

    /**
     * (internal) Sets the handler UUID.
     *
     * @param handlerUuid the handler UUID received from the server.
     * @return the {@link Builder} for chaining purposes.
     */
    Builder handlerUuid(int handlerUuid) {
      this.handlerUuid = handlerUuid;
      return this;
    }

    /**
     * (internal) Sets the function instance id.
     *
     * @param functionInstanceId the function instance ID received from the server.
     * @return the {@link Builder} for chaining purposes.
     */
    Builder functionInstanceId(String functionInstanceId) {
      this.functionInstanceId = functionInstanceId;
      return this;
    }

    /**
     * Builds the immutable {@link EventingFunction}.
     */
    public EventingFunction build() {
      return new EventingFunction(this);
    }

  }

  @Override
  public String toString() {
    return "EventingFunction{" +
      "name='" + name + '\'' +
      ", code='" + code + '\'' +
      ", sourceKeyspace=" + sourceKeyspace +
      ", metadataKeyspace=" + metadataKeyspace +
      ", settings=" + settings +
      ", version='" + version + '\'' +
      ", enforceSchema=" + enforceSchema +
      ", handlerUuid=" + handlerUuid +
      ", functionInstanceId='" + functionInstanceId + '\'' +
      ", bucketBindings=" + bucketBindings +
      ", urlBindings=" + urlBindings +
      ", constantBindings=" + constantBindings +
      '}';
  }
}

File: scala-client/src/main/scala/com/couchbase/client/scala/manager/eventing/EventingFunction.scala
/*
 * Copyright 2021 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.scala.manager.eventing

import com.couchbase.client.scala.Collection
import com.couchbase.client.scala.query.QueryScanConsistency

import scala.concurrent.duration.Duration

case class EventingFunction(
    name: String,
    code: String,
    sourceKeyspace: EventingFunctionKeyspace,
    metadataKeyspace: EventingFunctionKeyspace,
    settings: Option[EventingFunctionSettings] = None,
    version: Option[String] = None,
    enforceSchema: Option[Boolean] = None,
    handlerUuid: Option[Long] = None,
    functionInstanceId: Option[String] = None,
    bucketBindings: Option[Seq[EventingFunctionBucketBinding]] = None,
    urlBindings: Option[Seq[EventingFunctionUrlBinding]] = None,
    constantBindings: Option[Seq[EventingFunctionConstantBinding]] = None
) {}

object EventingFunction {
  def create(
      name: String,
      code: String,
      sourceKeyspace: EventingFunctionKeyspace,
      metadataKeyspace: EventingFunctionKeyspace
  ): EventingFunction = {
    EventingFunction(name, code, sourceKeyspace, metadataKeyspace)
  }
}

case class EventingFunctionBucketBinding(
    alias: String,
    name: EventingFunctionKeyspace,
    access: EventingFunctionBucketAccess
)

sealed trait EventingFunctionBucketAccess

object EventingFunctionBucketAccess {
  case object ReadOnly extends EventingFunctionBucketAccess

  case object ReadWrite extends EventingFunctionBucketAccess
}

case class EventingFunctionUrlBinding(
    hostname: String,
    alias: String,
    auth: EventingFunctionUrlAuth,
    allowCookies: Boolean,
    validateSslCertificate: Boolean
)

sealed trait EventingFunctionUrlAuth

object EventingFunctionUrlAuth {
  case object None extends EventingFunctionUrlAuth

  case class Basic(username: String, password: Option[String]) extends EventingFunctionUrlAuth

  case class Digest(username: String, password: Option[String]) extends EventingFunctionUrlAuth

  case class Bearer(key: String) extends EventingFunctionUrlAuth
}

case class EventingFunctionConstantBinding(alias: String, literal: String)

case class EventingFunctionSettings(
    cppWorkerThreadCount: Option[Long] = None,
    dcpStreamBoundary: Option[EventingFunctionDcpBoundary] = None,
    description: Option[String] = None,
    logLevel: Option[EventingFunctionLogLevel] = None,
    languageCompatibility: Option[EventingFunctionLanguageCompatibility] = None,
    executionTimeout: Option[Duration] = None,
    lcbInstCapacity: Option[Long] = None,
    lcbRetryCount: Option[Long] = None,
    lcbTimeout: Option[Duration] = None,
    queryConsistency: Option[QueryScanConsistency] = None,
    numTimerPartitions: Option[Long] = None,
    sockBatchSize: Option[Long] = None,
    tickDuration: Option[Duration] = None,
    timerContextSize: Option[Long] = None,
    userPrefix: Option[String] = None,
    bucketCacheSize: Option[Long] = None,
    bucketCacheAge: Option[Long] = None,
    curlMaxAllowedRespSize: Option[Long] = None,
    workerCount: Option[Long] = None,
    queryPrepareAll: Option[Boolean] = None,
    handlerHeaders: Option[Seq[String]] = None,
    handlerFooters: Option[Seq[String]] = None,
    enableAppLogRotation: Option[Boolean] = None,
    appLogDir: Option[String] = None,
    appLogMaxSize: Option[Long] = None,
    appLogMaxFiles: Option[Long] = None,
    checkpointInterval: Option[Duration] = None,
    processingStatus: Option[EventingFunctionProcessingStatus] = None,
    deploymentStatus: Option[EventingFunctionDeploymentStatus] = None
) {}

sealed trait EventingFunctionDcpBoundary {}

object EventingFunctionDcpBoundary {
  case object Everything extends EventingFunctionDcpBoundary

  case object FromNow extends EventingFunctionDcpBoundary
}

sealed trait EventingFunctionDeploymentStatus {}

object EventingFunctionDeploymentStatus {
  case object Deployed extends EventingFunctionDeploymentStatus

  case object Undeployed extends EventingFunctionDeploymentStatus
}

sealed trait EventingFunctionProcessingStatus

object EventingFunctionProcessingStatus {
  case object Running extends EventingFunctionProcessingStatus

  case object Paused extends EventingFunctionProcessingStatus
}

sealed trait EventingFunctionLogLevel

object EventingFunctionLogLevel {
  case object Info extends EventingFunctionLogLevel

  case object Error extends EventingFunctionLogLevel

  case object Warning extends EventingFunctionLogLevel

  case object Debug extends EventingFunctionLogLevel

  case object Trace extends EventingFunctionLogLevel
}

sealed trait EventingFunctionLanguageCompatibility {}

object EventingFunctionLanguageCompatibility {
  case object Version_6_0_0 extends EventingFunctionLanguageCompatibility

  case object Version_6_5_0 extends EventingFunctionLanguageCompatibility

  case object Version_6_6_2 extends EventingFunctionLanguageCompatibility

  case object Version_7_2_0 extends EventingFunctionLanguageCompatibility
}

case class EventingFunctionKeyspace(
    bucket: String,
    scope: Option[String] = None,
    collection: Option[String] = None
)

object EventingFunctionKeyspace {
  def createFrom(collection: Collection): EventingFunctionKeyspace = {
    EventingFunctionKeyspace(
      collection.bucketName,
      Some(collection.scopeName),
      Some(collection.name)
    )
  }
}

case class EventingStatus(numEventingNodes: Int, functions: Seq[EventingFunctionState])

case class EventingFunctionState(
    name: String,
    status: EventingFunctionStatus,
    numBootstrappingNodes: Int,
    numDeployedNodes: Int,
    deploymentStatus: EventingFunctionDeploymentStatus,
    processingStatus: EventingFunctionProcessingStatus
)

sealed trait EventingFunctionStatus

object EventingFunctionStatus {
  case object Undeployed extends EventingFunctionStatus

  case object Deploying extends EventingFunctionStatus

  case object Deployed extends EventingFunctionStatus

  case object Undeploying extends EventingFunctionStatus

  case object Paused extends EventingFunctionStatus

  case object Pausing extends EventingFunctionStatus
}
